### PR TITLE
Added means to read and patch relocations to .rodata

### PR DESCRIPTION
**Pull request by llasek-dt**
Updating the BPF program loading flow:
1. loading of all ELF sections,
2. creation of bpf maps,
3. relocations patching.

For this purpose, the MapsSectionLoader is generalized and renamed to SectionLoader.
ELF section reading is now decoupled from the creation of maps_config. Separate map is created for .rodata section.
Added means to patch .rodata relocations. Relocations patching is updated to check for R_BPF_64_64 relocation type.

Added `BPF_PRINTK(fmt, ...)` helper macro for `bpf_trace_printk()`.

Additionally, fixed the following issues:
* bpf object file overwrites by relocations patching,
* build issue related to llvm-dev package version.

### CODE REVIEW

File: bpf_generic/src/maps_loading.cpp [hunk 1]
```diff
@@ -122,99 +99,195 @@ bool loadMaps(maps_config& maps, BPFMapsWrapper& mapsWrapper) {
...
+			auto relSymbol = rel.getSymbol();
+			auto relSymbolAddr = relSymbol->getAddress();
+			if (!relSymbolAddr) {
+				LOG_ERROR("prog:{}+{}: unknown relocation addr", kprobeName, offs);
+				return false;
+			}
+
+			// (src_reg, imm) mapping ref. https://www.kernel.org/doc/html/latest/bpf/standardization/instruction-set.html#bit-immediate-instructions
+			if (sections.maps->containsSymbol(*relSymbol)) {
+				auto map = std::find_if(maps.begin(), maps.end(), [&relSymbolAddr](auto& map) { return map.elf_offset == *relSymbolAddr; });
```
Comment (joannaglowala) [inline bpf_generic/src/maps_loading.cpp:226]: Although it's not expected after containsSymbol() check, result of find_if should be verified anyway.

File: bpf_program/print.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#ifdef LEGACY_BPF
+#include "legacy/bpf_helpers.h"
+#else
+#include <bpf/bpf_helpers.h>
+#endif
+
+#if USE_RODATA
```
Comment (joannaglowala) [inline bpf_program/print.h:9]: How/where is it expected to be defined?

File: bpf_generic/src/maps_loading.cpp [hunk 3]
```diff
@@ -122,99 +99,195 @@ bool loadMaps(maps_config& maps, BPFMapsWrapper& mapsWrapper) {
...
-				continue;
+		// must copy the bpf prog body: llvm elf structs are mmap'd directly from .o file, and patching relocs would overwrite the file
+		auto [itBody, ok] = bpfPrograms.emplace(kprobeName, std::vector<char>(kprobeBody.begin(), kprobeBody.end()));
+		auto* bytecode = std::bit_cast<bpf_insn*>(itBody->second.data());
+
+		// read and patch kprobe relocations
+		for (auto& rel : kprobeRel.relocations()) {
+			auto relOffs = rel.getOffset();
+			if (relOffs >= itBody->second.size()) {
+				LOG_ERROR("prog:{}[{}]: invalid relo offset {}", kprobeName, itBody->second.size(), relOffs);
```
Comment (joannaglowala) [inline bpf_generic/src/maps_loading.cpp:203]: Shouldn't it stop the flow?

File: bpf_generic/src/maps_loading.cpp [hunk 4]
```diff
@@ -112,6 +83,12 @@ bool loadMaps(maps_config& maps, BPFMapsWrapper& mapsWrapper) {
 		}
 		map.fd = fd;
 
+		// init (memcpy) rodata
+		if (map.name == ".rodata" && rodataSec) {
+			int zero = 0;
+			mapsWrapper.updateElement(fd, &zero, rodataSec->getContents()->data());
```
Comment (joannaglowala) [inline bpf_generic/src/maps_loading.cpp]: Return value not checked.

Comment (joannaglowala) [inline bpf_generic/src/maps_loading.cpp]: rodataSec->getContents() not checked.

File: bpf_generic/src/maps_loading.cpp [hunk 5]
```diff
@@ -122,99 +99,195 @@ bool loadMaps(maps_config& maps, BPFMapsWrapper& mapsWrapper) {
...
-    }
-	return symbols;
+	}
+	return true;
 }
 
-maps_config MapsSectionLoader::copyElfMapsDataToMapsConfig(const MapsSymbols& symbols, std::size_t map_sz_copy) const {
+maps_config SectionLoader::getMapsConfig() {
+	auto mapsBody = sections.maps->getContents();
+	const std::size_t map_sz_elf = mapsBody->size() / mapsRelSymOffsToName.size();
```
Comment (joannaglowala) [inline bpf_generic/src/maps_loading.cpp:250]: Not sure how (im)probable it is, but mapsRelSymOffsToName.size() is not checked for zero before division.

**Review by joannaglowala — CHANGES_REQUESTED** [review]

File: bpf_generic/src/maps_loading.cpp (hunk 4 above)
Comment (llasek-dt) [reply bpf_generic/src/maps_loading.cpp]: It's checked upon loading sections: https://github.com/dynatrace-oss/nettracer-bpf/pull/61/changes#diff-199b2d481972b101b90d7a4f184dd3dc33bf894d82000844b1ff97ee247bdc2eR164-R175

File: bpf_generic/src/maps_loading.cpp (hunk 5 above)
Comment (llasek-dt) [reply bpf_generic/src/maps_loading.cpp:250]: Adding test for map empty in `loadSections()`: https://github.com/dynatrace-oss/nettracer-bpf/pull/61/changes#diff-199b2d481972b101b90d7a4f184dd3dc33bf894d82000844b1ff97ee247bdc2eR178

File: bpf_program/print.h (hunk 2 above)
Comment (llasek-dt) [reply bpf_program/print.h:9]: https://github.com/dynatrace-oss/nettracer-bpf/pull/61/changes#diff-887305cc0451e0c9971af6923bbc94a33611ff9947bcf5be192ea2b508b2df45R38-R64

File: bpf_generic/src/maps_loading.cpp (hunk 1 above)
Comment (llasek-dt) [reply bpf_generic/src/maps_loading.cpp:226]: This can't happen b/c all rel symbols to `maps` section are added to `maps` map... but okay, this happens once on init.

**Review by joannaglowala — APPROVED** [review]
